### PR TITLE
feat: add early stopping to Trainer.train

### LINKED ISSUE #216: Add Early stopping
Early stopping is a form of regularization used to avoid overfitting the training dataset. Early stopping keeps track of the validation loss; if the loss stops decreasing for several epochs in a row the training stops. We should have something like [this](https://github.com/Bjarten/early-stopping-pytorch).

### PR DESCRIPTION
**Pull request by DaniBodor**
This PR changes:
- add an adaptation of [this module](https://github.com/Bjarten/early-stopping-pytorch/blob/master/pytorchtools.py) for early stopping
  - I added gap functionality (check for difference between val loss and train loss)
  - other changes have to do with optional model saving, verbosity, status feedback messages, etc.
- change model saving and naming logic of `Trainer.train()` method
- improve docstrings & log info of `Trainer.train()` method
- add a unit test for earlystopping module
  - Not sure how useful this unit test is...

closes #216

### CODE REVIEW

**Comment by DaniBodor** [comment]
The unit test I created only tests the basic logic of the early stopper using dummy loss values without training anything. 

Ideally we would also have a functional/integrated test to run on an actually trained model to check that losses are passed to and processed by `EarlyStopper` as intended? However, I am not sure how to implement this in a way that forces a training that will trigger the early stopping criteria in a controlled (and timely) manner.

**Review by gcroci2 — APPROVED** [review]

**Comment by gcroci2** [comment]
> The unit test I created only tests the basic logic of the early stopper using dummy loss values without training anything.
> 
> Ideally we would also have a functional/integrated test to run on an actually trained model to check that losses are passed to and processed by `EarlyStopper` as intended? However, I am not sure how to implement this in a way that forces a training that will trigger the early stopping criteria in a controlled (and timely) manner. Any idea how to do this?

The only thing that comes to my mind is to set up an "high" patience number with a very small dataset (like 20, 30, 50, depending on how fast the training remains) and then assess that the model saved is at least after the patience epoch number.

File: deeprankcore/utils/earlystopping.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    def __init__( # pylint: disable=too-many-arguments
+        self,
+        patience: int = 10,
+        delta: Optional[float] = None,
+        maxgap: Optional[float] = None,
+        verbose: bool = True,
+        path: str = 'checkpoint.pt',
+        trace_func: Callable = print,
+    ):
+        """Terminate training if validation loss doesn't improve after a given patience or if a maximum gap between validation and training loss is reached.
```
Comment (DaniBodor) [inline deeprankcore/utils/earlystopping.py:16]:
```suggestion
        """
        Terminate training if validation loss doesn't improve after a given patience or if a maximum gap between validation and training loss is reached.
```

**Review by gcroci2 — APPROVED** [review]

File: deeprankcore/utils/earlystopping.py [hunk 2]
```diff
@@ -1,64 +1,53 @@
...
 
         Args:
             patience (int): How long to wait after last time validation loss improved.
-                            Default: 10
+                Default: 10
             
-            delta (float, optional): Minimum change in the monitored quantity to qualify as an improvement.
-                            Default: None
+            delta (float): Minimum change required to reset the early stopping counter.
+                Default: None
```
Comment (gcroci2) [inline deeprankcore/utils/earlystopping.py:21]:
```suggestion
                Default: 0
```

**Review by gcroci2 — APPROVED** [review]
